### DISCUSSION TITLE
Can closing by clicking the X on Windows be handled?

### DISCUSSION THREAD

**Discussion by David-H-Bolton** [tui-cs/Terminal.Gui#1597] [Q&A] [answered]
It seems to do a complete close down without allowing a chance for any kind of CloseDown method be called or offer a closedown dialog- "Are you sure you want to close?". I'm new to Terminal.gui so my apologies if I've missed anything obvious but I looked through the documentation and didn't see anything.

**Comment by BDisp** [comment]
No, sorry.

**Comment by tznind** [comment, marked as answer]
Do you mean the X on the actual terminal window itself for example:

![image](https://user-images.githubusercontent.com/31306100/154471427-5b86877e-6889-4f33-8d38-acd76d86e6ad.png)

**Reply by BDisp** [reply]
Using as MDI it is possible to control the closing in each child but not on the console itself.

**Reply by tznind** [reply]
Just wanted to check that the X you were talking about @David-H-Bolton  was the console itself and not something within Terminal.Gui.  It is possible to do cleanup on application exit by using the `AppDomain.CurrentDomain.ProcessExit` event but you cannot cancel it.  I also suspect that any blocking/long running operation at that stage is likely to go badly.

I checked the following code on PowerShell and Cmd and both worked as intended (when you close the console you get the "heya.txt" still appearing.

```csharp
// See https://aka.ms/new-console-template for more information
Console.WriteLine("Hello, World!");

AppDomain.CurrentDomain.ProcessExit += new EventHandler(CurrentDomain_ProcessExit);

while(true)
{

}

static void CurrentDomain_ProcessExit(object sender, EventArgs e)
{
    File.WriteAllText("heya.txt", "it closed on us!");
}
```

**Reply by David-H-Bolton** [reply]
Yes I meant the top right X.

**Reply by David-H-Bolton** [reply]
Ah, that's what I was looking for.  AppDomain.CurrentDomain.ProcessExit . Thank you. It means I can always guarantee that a SaveState sort of method can be called.

**Reply by tznind** [reply]
Yeah unfortunately that is outside of the control of the C# program that is running (i.e. your app).  It may be that you could use the above approach to persist the application state (which windows were open, what values fields had etc) so that on the next startup you can resume that state.

Edit: Just saw your reply :) sounds good.

**Reply by BDisp** [reply]
> I also suspect that any blocking/long running operation at that stage is likely to go badly.

@tznind, if it's give time enough may be propagated an exiting event through the `Application.Top`, with this event `AppDomain.CurrentDomain.ProcessExit` setted up, to the all `toplevels`. Is it worth it?

**Reply by tznind** [reply]
> > I also suspect that any blocking/long running operation at that stage is likely to go badly.
> 
> @tznind, if it's give time enough may be propagated an exiting event through the `Application.Top`, with this event `AppDomain.CurrentDomain.ProcessExit` setted up, to the all `toplevels`. Is it worth it?

Hmn... I don't think theres much point wrapping that event and renaming/exposing it in Terminal.Gui since it is part of core C# and is quite easy to just use directly.
